plugins/documents/app/controllers/documents_controller.rb

```diff
@@ -0,0 +1,63 @@
+require 'filemagic'
+
+class DocumentsController < ApplicationController
+  before_filter -> { require_plugin_enabled FoodsoftDocuments }
+
+  def index
+    if params["sort"]
+      sort = case params["sort"]
+               when "name" then "name"
+               when "created_at" then "created_at"
+               when "name_reverse" then "name DESC"
+               when "created_at_reverse" then "created_at DESC"
+             end
+    else
+      sort = "name"
+    end
+
+    @documents = Document.page(params[:page]).per(@per_page).order(sort)
+  end
+
+  def new
+    @document = Document.new
+  end
+
+  def create
+    @document = Document.new
+    @document.data = params[:document][:data].read
+    @document.mime = FileMagic.new(FileMagic::MAGIC_MIME).buffer(@document.data)
+    if params[:document][:name] == ''
+      name = params[:document][:data].original_filename
+      name = File.basename(name)
+      @document.name = name.gsub(/[^\w\.\-]/, '_')
```

Comment:
blacklisting as bad practice and whitelisting all unicode characters gets ugly very quickly. What's the problem with storing mad filenames? The only place where it could be problematic is when the browser reads the filename from the HTTP header and when the user uses a browser which is not able to to the without security impact (s)he might have even bigger problems than that!